src/test/java/DirectionsServletTest.java

```diff
@@ -0,0 +1,175 @@
+    when(directionsClientFactory.getDirectionsClient(Mockito.any())).thenReturn(directionsClient);
+  }
+
+  @Before
+  public void init() throws IOException {
+    request = Mockito.mock(HttpServletRequest.class);
+    response = Mockito.mock(HttpServletResponse.class);
+
+    // Writer used in get/post requests to capture HTTP response values
+    stringWriter = new StringWriter();
+    printWriter = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(printWriter);
+  }
+
+  @After
+  public void clear() {
+    // Dump contents after each test
+    stringWriter.getBuffer().setLength(0);
+  }
+
+  @Test
+  public void windsorToMontrealWithWaypoints() {
+    // Get optimized route of travel from Windsor, ON to Montreal, QC with waypoints
+    // Markham, ON,
+    // Waterloo, ON and Quebec City, QC in between.
+    // Should return four legs: Windsor, ON -> Waterloo, ON -> Markham, ON -> Quebec
+    // City, QC ->
+    // Montreal, QC.
+    try {
+      when(directionsClient.getDirections(
+              Mockito.anyString(), Mockito.anyString(), Mockito.any(String[].class)))
+          .thenReturn(windsorToMontrealWithWaypoints);
+    } catch (ApiException | InterruptedException | IOException e) {
+      e.printStackTrace();
+    }
+    servlet.doGet(request, response);
+    printWriter.flush();
+    // Replace unicode and escaped characters
+    String actual =
+        stringWriter
+            .toString()
+            .replace("\\u003e", ">")
+            .replace("\\u003d", "=")
+            .replace("\\\"", "\"");
+    Assert.assertTrue(actual.contains(WINDSOR_TO_MONTREAL_WITH_WAYPOINTS_JSON));
+  }
+
+  @Test
+  public void montrealToMontrealWithWaypoints() {
+    // Get optimized route of travel from Montreal, QC to Montreal, QC with
+    // waypoints Windsor, ON,
+    // Waterloo, ON and Kitchener, ON in between.
+    // Should return four legs: Montreal, QC -> Kitchener, ON -> Waterloo, ON ->
+    // Windsor, ON ->
+    // Montreal, QC.
+    try {
+      when(directionsClient.getDirections(
+              Mockito.anyString(), Mockito.anyString(), Mockito.any(String[].class)))
+          .thenReturn(montrealToMontrealWithWaypoints);
+    } catch (ApiException | InterruptedException | IOException e) {
+      e.printStackTrace();
+    }
+    servlet.doGet(request, response);
+    printWriter.flush();
+    // Replace unicode and escaped characters
+    String actual =
+        stringWriter
+            .toString()
+            .replace("\\u003e", ">")
+            .replace("\\u003d", "=")
+            .replace("\\\"", "\"");
+    Assert.assertTrue(actual.contains(MONTREAL_TO_MONTREAL_WITH_WAYPOINTS_JSON));
+  }
+
+  @Test
+  public void montrealToWaterlooNoWaypoints() {
+    // Get optimized route of travel from Montreal, QC to Waterloo, ON with no
+    // waypoints in between.
+    // Should return only one leg from Montreal, QC to Waterloo, ON.
```

Comment:
An advantage of simplifying the test data is that it would allow you to remove these comments as readers would be able to understand the input/expected output from the data itself.

Reply:
Great point! Just to confirm, are comments optional for public tests?

Reply:
Yes. Apart from Javadoc (and equivalents in other languages), comments are always optional. If your code is clear enough and you can enforce everything you need using the language, there's no need for comments. If you want to take a look at the tests for the Guava project https://github.com/google/guava/tree/master/guava-tests/test/com/google/common, you will see that it is very common for tests not to have any comment.

Reply:
This is a great example, will definitely refer to this when in doubt. Thanks!

Reply:
Just to confirm - is there an outstanding request to simplify the test data, or is this comment resolved?

Reply:
Yes there is an outstanding request #81 to simplify the test data so the comments would still be needed for now. Here's the thread to the related review comment https://github.com/googleinterns/step7-2020/pull/63#discussion_r455368196.

Reply:
`RESOLVED`